python3Packages.jaxlib: fix #282184 and migrate to cuda redist packages

samuela opened the pull request:
## Description of changes

This change snowballed to cover a number of improvements:
1. Fix https://github.com/NixOS/nixpkgs/issues/282184. The `--config=cuda` flag was lost at some point, disabling CUDA builds even with `cudaSupport = true`.
2. Migrate to using the CUDA redist packages instead of cudatoolkit.
3. Unify stdenv behind `effectiveStdenv` following a pattern that has precedent in OpenCV's derivation and was recommended in https://github.com/NixOS/nixpkgs/pull/288857#discussion_r1490003016.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

samuela commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -204,40 +213,60 @@ let
+      # TODO: test whether we still need this
+      ''
+        mkdir dummy-ldconfig
+        echo "#!${backendStdenv.shell}" > dummy-ldconfig/ldconfig
```
@SomeoneSerge I'm assuming that this ought to be `backendStdenv.shell` instead of `stdenv.shell` but I'm not really sure. There are also a number of other instances where `stdenv` is used in the derivation... not sure which of those, if any, we should switch over?

SomeoneSerge replied:
1. > assuming that this ought to be backendStdenv.shell instead of stdenv.shell

   in principle, this shouldn't matter, I think they refer to the same shell in practice
2. > There are also a number of other instances where stdenv is used in the derivation

   ...would be nice to be consistent, e.g. the way our opencv derivation does it (defines the `effectiveStdenv`, and hides the old `stdenv` with the new `throw` definition)
3. ...but would also be nice to kill the `backendStdenv` off

samuela replied:
ahhh tough choices! ok i'll plan on going the `effectiveStdenv` route, if that sounds acceptable to you? we can always remove `backendStdenv` later once the path to doing so is clear

SomeoneSerge replied:
Yes, when we remove `backendStdenv` we remove it treewide. It's not clear to me how to do that yet

samuela replied:
done in https://github.com/NixOS/nixpkgs/pull/288857/commits/c0e2764852da93be1ee5b6f1d199bcced94385c2

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -204,40 +213,60 @@ let
 
-    GCC_HOST_COMPILER_PREFIX = lib.optionalString cudaSupport "${cudatoolkit_cc_joined}/bin";
-    GCC_HOST_COMPILER_PATH = lib.optionalString cudaSupport "${cudatoolkit_cc_joined}/bin/gcc";
+    GCC_HOST_COMPILER_PREFIX = lib.optionalString cudaSupport "${backend_cc_joined}/bin";
+    GCC_HOST_COMPILER_PATH = lib.optionalString cudaSupport "${backend_cc_joined}/bin/gcc";
```
Since touching anyway, consider maybe adding the `targetPrefix` (grep nixpkgs for examples). Ok to ignore bc out of scope

samuela replied:
Roger that, I'll take a stab at this once I get the build working

samuela replied:
i genuinely appreciate this pointer, this PR has already gotten quite large so i propose that we leave that for future work in the interest of keeping PRs small. Created https://github.com/NixOS/nixpkgs/issues/289427 to track

samuela commented:
Currently blocked on https://github.com/google/jax/issues/19814

samuela commented:
@SomeoneSerge @GaetanLepage @ndl This fixes the CUDA jaxlib build, and is now ready for review. Let me know if you have any feedback!

Only remaining TODO AFAIK is to get the aarch64-linux fetchAttrs hash.

samuela commented:
I don't currently have access to an aarch64-linux machine to fetch the hash with. Anyone who might be able to help out? If not, I propose that we drop support for the time being.

GaetanLepage commented:
Result of `nixpkgs-review pr 288857` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.dm-sonnet</li>
    <li>python311Packages.dm-sonnet.dist</li>
    <li>python311Packages.elegy</li>
    <li>python311Packages.elegy.dist</li>
    <li>python311Packages.tensorflow-datasets</li>
    <li>python311Packages.tensorflow-datasets.dist</li>
    <li>python311Packages.treeo</li>
    <li>python311Packages.treeo.dist</li>
    <li>python311Packages.treex</li>
    <li>python311Packages.treex.dist</li>
  </ul>
</details>
<details>
  <summary>26 packages failed to build:</summary>
  <ul>
    <li>libretranslate (python311Packages.libretranslate)</li>
    <li>libretranslate.dist (python311Packages.libretranslate.dist)</li>
    <li>python311Packages.aeppl</li>
    <li>python311Packages.aeppl.dist</li>
    <li>python311Packages.aesara</li>
    <li>python311Packages.aesara.dist</li>
    <li>python311Packages.augmax</li>
    <li>python311Packages.augmax.dist</li>
    <li>python311Packages.coffea</li>
    <li>python311Packages.coffea.dist</li>
    <li>python311Packages.dalle-mini</li>
    <li>python311Packages.dalle-mini.dist</li>
    <li>python311Packages.distrax</li>
    <li>python311Packages.distrax.dist</li>
    <li>python311Packages.dm-haiku</li>
    <li>python311Packages.dm-haiku.dist</li>
    <li>python311Packages.flax</li>
    <li>python311Packages.flax.dist</li>
    <li>python311Packages.objax</li>
    <li>python311Packages.objax.dist</li>
    <li>python311Packages.rlax</li>
    <li>python311Packages.rlax.dist</li>
    <li>python311Packages.vqgan-jax</li>
    <li>python311Packages.vqgan-jax.dist</li>
    <li>tts</li>
    <li>tts.dist</li>
  </ul>
</details>
<details>
  <summary>63 packages built:</summary>
  <ul>
    <li>python311Packages.argos-translate-files</li>
    <li>python311Packages.argos-translate-files.dist</li>
    <li>python311Packages.argostranslate</li>
    <li>python311Packages.argostranslate.dist</li>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.awkward</li>
    <li>python311Packages.awkward.dist</li>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
    <li>python311Packages.blackjax</li>
    <li>python311Packages.blackjax.dist</li>
    <li>python311Packages.chex</li>
    <li>python311Packages.chex.dist</li>
    <li>python311Packages.correctionlib</li>
    <li>python311Packages.correctionlib.dist</li>
    <li>python311Packages.ctranslate2</li>
    <li>python311Packages.ctranslate2.dist</li>
    <li>python311Packages.dask-awkward</li>
    <li>python311Packages.dask-awkward.dist</li>
    <li>python311Packages.equinox</li>
    <li>python311Packages.equinox.dist</li>
    <li>python311Packages.faster-whisper</li>
    <li>python311Packages.faster-whisper.dist</li>
    <li>python311Packages.gymnasium</li>
    <li>python311Packages.gymnasium.dist</li>
    <li>python311Packages.jax</li>
    <li>python311Packages.jax.dist</li>
    <li>python311Packages.jaxlib (python311Packages.jaxlib-build ,python311Packages.jaxlibWithoutCuda)</li>
    <li>python311Packages.jaxlib.dist (python311Packages.jaxlib-build.dist ,python311Packages.jaxlibWithoutCuda.dist)</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.jaxlibWithCuda.dist</li>
    <li>python311Packages.jaxopt</li>
    <li>python311Packages.jaxopt.dist</li>
    <li>python311Packages.jmp</li>
    <li>python311Packages.jmp.dist</li>
    <li>python311Packages.mplhep</li>
    <li>python311Packages.mplhep.dist</li>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
    <li>python311Packages.optax</li>
    <li>python311Packages.optax.dist</li>
    <li>python311Packages.optax.testsout</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
    <li>python311Packages.pytensor</li>
    <li>python311Packages.pytensor.dist</li>
    <li>python311Packages.skrl</li>
    <li>python311Packages.skrl.dist</li>
    <li>python311Packages.tensorflow-bin</li>
    <li>python311Packages.tensorflow-bin.dist</li>
    <li>python311Packages.torchrl</li>
    <li>python311Packages.torchrl.dist</li>
    <li>python311Packages.translatehtml</li>
    <li>python311Packages.translatehtml.dist</li>
    <li>python311Packages.uproot</li>
    <li>python311Packages.uproot.dist</li>
    <li>python311Packages.vector</li>
    <li>python311Packages.vector.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
    <li>wyoming-faster-whisper</li>
    <li>wyoming-faster-whisper.dist</li>
  </ul>
</details>

samuela commented:
> Amazing work ! Thanks so much @samuela !
Hash for aarch64-linux is: sha256-NehnpA4m+Fynvh0S6WKy/v9ab81487NE9ahvbS70wjY=

Thanks @GaetanLepage! Just pushed a commit with the aarch64-linux fetchAttrs hash, but is otherwise identical to the previous one

GaetanLepage commented:
> but is otherwise identical to the previous one

This is not suprising as you have just changed cuda-related stuff right ?

GaetanLepage commented:
Result of `nixpkgs-review pr 288857` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.dm-sonnet</li>
    <li>python311Packages.dm-sonnet.dist</li>
    <li>python311Packages.elegy</li>
    <li>python311Packages.elegy.dist</li>
    <li>python311Packages.tensorflow-datasets</li>
    <li>python311Packages.tensorflow-datasets.dist</li>
    <li>python311Packages.treeo</li>
    <li>python311Packages.treeo.dist</li>
    <li>python311Packages.treex</li>
    <li>python311Packages.treex.dist</li>
  </ul>
</details>
<details>
  <summary>26 packages failed to build:</summary>
  <ul>
    <li>libretranslate (python311Packages.libretranslate)</li>
    <li>libretranslate.dist (python311Packages.libretranslate.dist)</li>
    <li>python311Packages.aeppl</li>
    <li>python311Packages.aeppl.dist</li>
    <li>python311Packages.aesara</li>
    <li>python311Packages.aesara.dist</li>
    <li>python311Packages.augmax</li>
    <li>python311Packages.augmax.dist</li>
    <li>python311Packages.coffea</li>
    <li>python311Packages.coffea.dist</li>
    <li>python311Packages.dalle-mini</li>
    <li>python311Packages.dalle-mini.dist</li>
    <li>python311Packages.distrax</li>
    <li>python311Packages.distrax.dist</li>
    <li>python311Packages.dm-haiku</li>
    <li>python311Packages.dm-haiku.dist</li>
    <li>python311Packages.flax</li>
    <li>python311Packages.flax.dist</li>
    <li>python311Packages.objax</li>
    <li>python311Packages.objax.dist</li>
    <li>python311Packages.rlax</li>
    <li>python311Packages.rlax.dist</li>
    <li>python311Packages.vqgan-jax</li>
    <li>python311Packages.vqgan-jax.dist</li>
    <li>tts</li>
    <li>tts.dist</li>
  </ul>
</details>
<details>
  <summary>63 packages built:</summary>
  <ul>
    <li>python311Packages.argos-translate-files</li>
    <li>python311Packages.argos-translate-files.dist</li>
    <li>python311Packages.argostranslate</li>
    <li>python311Packages.argostranslate.dist</li>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.awkward</li>
    <li>python311Packages.awkward.dist</li>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
    <li>python311Packages.blackjax</li>
    <li>python311Packages.blackjax.dist</li>
    <li>python311Packages.chex</li>
    <li>python311Packages.chex.dist</li>
    <li>python311Packages.correctionlib</li>
    <li>python311Packages.correctionlib.dist</li>
    <li>python311Packages.ctranslate2</li>
    <li>python311Packages.ctranslate2.dist</li>
    <li>python311Packages.dask-awkward</li>
    <li>python311Packages.dask-awkward.dist</li>
    <li>python311Packages.equinox</li>
    <li>python311Packages.equinox.dist</li>
    <li>python311Packages.faster-whisper</li>
    <li>python311Packages.faster-whisper.dist</li>
    <li>python311Packages.gymnasium</li>
    <li>python311Packages.gymnasium.dist</li>
    <li>python311Packages.jax</li>
    <li>python311Packages.jax.dist</li>
    <li>python311Packages.jaxlib (python311Packages.jaxlib-build ,python311Packages.jaxlibWithoutCuda)</li>
    <li>python311Packages.jaxlib.dist (python311Packages.jaxlib-build.dist ,python311Packages.jaxlibWithoutCuda.dist)</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.jaxlibWithCuda.dist</li>
    <li>python311Packages.jaxopt</li>
    <li>python311Packages.jaxopt.dist</li>
    <li>python311Packages.jmp</li>
    <li>python311Packages.jmp.dist</li>
    <li>python311Packages.mplhep</li>
    <li>python311Packages.mplhep.dist</li>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
    <li>python311Packages.optax</li>
    <li>python311Packages.optax.dist</li>
    <li>python311Packages.optax.testsout</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
    <li>python311Packages.pytensor</li>
    <li>python311Packages.pytensor.dist</li>
    <li>python311Packages.skrl</li>
    <li>python311Packages.skrl.dist</li>
    <li>python311Packages.tensorflow-bin</li>
    <li>python311Packages.tensorflow-bin.dist</li>
    <li>python311Packages.torchrl</li>
    <li>python311Packages.torchrl.dist</li>
    <li>python311Packages.translatehtml</li>
    <li>python311Packages.translatehtml.dist</li>
    <li>python311Packages.uproot</li>
    <li>python311Packages.uproot.dist</li>
    <li>python311Packages.vector</li>
    <li>python311Packages.vector.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
    <li>wyoming-faster-whisper</li>
    <li>wyoming-faster-whisper.dist</li>
  </ul>
</details>

samuela commented:
> This is not suprising as you have just changed cuda-related stuff right ?

Yup! just wanted to clarify to ease reviewing since I was force pushing

GaetanLepage commented:
Result of `nixpkgs-review pr 288857` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.aeppl</li>
    <li>python311Packages.aeppl.dist</li>
    <li>python311Packages.aesara</li>
    <li>python311Packages.aesara.dist</li>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
    <li>python311Packages.pytensor</li>
    <li>python311Packages.pytensor.dist</li>
    <li>python311Packages.treeo</li>
    <li>python311Packages.treeo.dist</li>
  </ul>
</details>
<details>
  <summary>24 packages failed to build:</summary>
  <ul>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.augmax</li>
    <li>python311Packages.augmax.dist</li>
    <li>python311Packages.awkward</li>
    <li>python311Packages.awkward.dist</li>
    <li>python311Packages.blackjax</li>
    <li>python311Packages.blackjax.dist</li>
    <li>python311Packages.coffea</li>
    <li>python311Packages.coffea.dist</li>
    <li>python311Packages.correctionlib</li>
    <li>python311Packages.correctionlib.dist</li>
    <li>python311Packages.dask-awkward</li>
    <li>python311Packages.dask-awkward.dist</li>
    <li>python311Packages.mplhep</li>
    <li>python311Packages.mplhep.dist</li>
    <li>python311Packages.objax</li>
    <li>python311Packages.objax.dist</li>
    <li>python311Packages.torchrl</li>
    <li>python311Packages.torchrl.dist</li>
    <li>python311Packages.uproot</li>
    <li>python311Packages.uproot.dist</li>
    <li>python311Packages.vector</li>
    <li>python311Packages.vector.dist</li>
  </ul>
</details>
<details>
  <summary>21 packages built:</summary>
  <ul>
    <li>python311Packages.chex</li>
    <li>python311Packages.chex.dist</li>
    <li>python311Packages.equinox</li>
    <li>python311Packages.equinox.dist</li>
    <li>python311Packages.gymnasium</li>
    <li>python311Packages.gymnasium.dist</li>
    <li>python311Packages.jax</li>
    <li>python311Packages.jax.dist</li>
    <li>python311Packages.jaxlib (python311Packages.jaxlib-build ,python311Packages.jaxlibWithoutCuda)</li>
    <li>python311Packages.jaxlib.dist (python311Packages.jaxlib-build.dist ,python311Packages.jaxlibWithoutCuda.dist)</li>
    <li>python311Packages.jaxopt</li>
    <li>python311Packages.jaxopt.dist</li>
    <li>python311Packages.jmp</li>
    <li>python311Packages.jmp.dist</li>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
    <li>python311Packages.optax</li>
    <li>python311Packages.optax.dist</li>
    <li>python311Packages.optax.testsout</li>
    <li>python311Packages.skrl</li>
    <li>python311Packages.skrl.dist</li>
  </ul>
</details>

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -137,8 +169,43 @@ let
+    };
+
+    patches = [
+      # See https://github.com/google/jax/issues/19814#issuecomment-1945141259.
```
Nit: I forget this myself but a link preview of sorts would be nice

samuela replied:
done in 833baeb

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -204,50 +269,80 @@ let
-      CFG
-    '';
+    preConfigure =
+      # Dummy ldconfig to work around "Can't open cache file /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/etc/ld.so.cache" error
```
Replace the hash with `eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee`

samuela replied:
i opted for `/nix/store/<hash>-...` to save a few electrons in 833baeb. lmk if that works for you!

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -302,13 +397,13 @@ let
-    if stdenv.hostPlatform.isLinux then
+    if effectiveStdenv.hostPlatform.isLinux then
       "manylinux2014_${arch}"
-    else if stdenv.system == "x86_64-darwin" then
+    else if effectiveStdenv.system == "x86_64-darwin" then
       "macosx_10_9_${arch}"
-    else if stdenv.system == "aarch64-darwin" then
+    else if effectiveStdenv.system == "aarch64-darwin" then
       "macosx_11_0_${arch}"
-    else throw "Unsupported target platform: ${stdenv.hostPlatform}";
+    else throw "Unsupported target platform: ${effectiveStdenv.hostPlatform}";
```
(TODO for myself: look into and maybe open a separate issue)

Out of scope, but I didn't notice this before and it looks suspicious that a downstream derivation is computing the python "platform tags", and the more so that it uses `throw`. I wonder the nixpkgs python machinery is already implementing this logic elsewhere

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -319,20 +414,18 @@ buildPythonPackage {
 
     find $out -type f \( -name '*.so' -or -name '*.so.*' \) | while read lib; do
-      addOpenGLRunpath "$lib"
-      patchelf --set-rpath "${cudatoolkit}/lib:${cudatoolkit.lib}/lib:${cudnn}/lib:${nccl}/lib:$(patchelf --print-rpath "$lib")" "$lib"
+      patchelf --set-rpath "${cuda_libs_joined}/lib:${cudnn}/lib:${nccl}/lib:$(patchelf --print-rpath "$lib")" "$lib"
```
Out of scope/TODO: `find ... -print0` and `IFS= read -r -d $'\0'` or how does one spell this nonsense

SomeoneSerge replied:
Also `--add-rpath` like in the other PR is an option

samuela replied:
done in https://github.com/NixOS/nixpkgs/pull/288857/commits/1e239b3ae2112a378b1697444410109b45fa53f6

samuela commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -137,8 +169,43 @@ let
+    if effectiveStdenv.hostPlatform.isLinux && effectiveStdenv.hostPlatform.linuxArch == "arm64" then "aarch64"
+    else effectiveStdenv.hostPlatform.linuxArch;
+
+  xla = effectiveStdenv.mkDerivation {
```
It additionally applies a patch from upstream that's necessary to fix the build and calls patchShebangs on build scripts. Perhaps there's a more lightweight way of expressing this?

SomeoneSerge replied:
I guess that's fine, I'm just thinking this is maybe a bit misleading in that the output isn't actually the (built) xla. Maybe we adjust the pname accordingly so it shows up nicely in `nom` and in `nix-tree`. I was thinking of something like `"xla-src"`, mimicking https://github.com/NixOS/nixpkgs/blob/83660f917bfdaa82a8a6344c82286ac8ad616378/pkgs/build-support/fetchgithub/default.nix#L3

Anyway, this is yak shaving...

samuela replied:
ah good point. done in 6e0570c

GaetanLepage commented:
Result of `nixpkgs-review pr 288857` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.dm-sonnet</li>
    <li>python311Packages.dm-sonnet.dist</li>
    <li>python311Packages.elegy</li>
    <li>python311Packages.elegy.dist</li>
    <li>python311Packages.tensorflow-datasets</li>
    <li>python311Packages.tensorflow-datasets.dist</li>
    <li>python311Packages.treeo</li>
    <li>python311Packages.treeo.dist</li>
    <li>python311Packages.treex</li>
    <li>python311Packages.treex.dist</li>
  </ul>
</details>
<details>
  <summary>24 packages failed to build:</summary>
  <ul>
    <li>libretranslate (python311Packages.libretranslate)</li>
    <li>libretranslate.dist (python311Packages.libretranslate.dist)</li>
    <li>python311Packages.aeppl</li>
    <li>python311Packages.aeppl.dist</li>
    <li>python311Packages.aesara</li>
    <li>python311Packages.aesara.dist</li>
    <li>python311Packages.augmax</li>
    <li>python311Packages.augmax.dist</li>
    <li>python311Packages.dalle-mini</li>
    <li>python311Packages.dalle-mini.dist</li>
    <li>python311Packages.distrax</li>
    <li>python311Packages.distrax.dist</li>
    <li>python311Packages.dm-haiku</li>
    <li>python311Packages.dm-haiku.dist</li>
    <li>python311Packages.flax</li>
    <li>python311Packages.flax.dist</li>
    <li>python311Packages.objax</li>
    <li>python311Packages.objax.dist</li>
    <li>python311Packages.rlax</li>
    <li>python311Packages.rlax.dist</li>
    <li>python311Packages.vqgan-jax</li>
    <li>python311Packages.vqgan-jax.dist</li>
    <li>tts</li>
    <li>tts.dist</li>
  </ul>
</details>
<details>
  <summary>65 packages built:</summary>
  <ul>
    <li>python311Packages.argos-translate-files</li>
    <li>python311Packages.argos-translate-files.dist</li>
    <li>python311Packages.argostranslate</li>
    <li>python311Packages.argostranslate.dist</li>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.awkward</li>
    <li>python311Packages.awkward.dist</li>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
    <li>python311Packages.blackjax</li>
    <li>python311Packages.blackjax.dist</li>
    <li>python311Packages.chex</li>
    <li>python311Packages.chex.dist</li>
    <li>python311Packages.coffea</li>
    <li>python311Packages.coffea.dist</li>
    <li>python311Packages.correctionlib</li>
    <li>python311Packages.correctionlib.dist</li>
    <li>python311Packages.ctranslate2</li>
    <li>python311Packages.ctranslate2.dist</li>
    <li>python311Packages.dask-awkward</li>
    <li>python311Packages.dask-awkward.dist</li>
    <li>python311Packages.equinox</li>
    <li>python311Packages.equinox.dist</li>
    <li>python311Packages.faster-whisper</li>
    <li>python311Packages.faster-whisper.dist</li>
    <li>python311Packages.gymnasium</li>
    <li>python311Packages.gymnasium.dist</li>
    <li>python311Packages.jax</li>
    <li>python311Packages.jax.dist</li>
    <li>python311Packages.jaxlib (python311Packages.jaxlib-build ,python311Packages.jaxlibWithoutCuda)</li>
    <li>python311Packages.jaxlib.dist (python311Packages.jaxlib-build.dist ,python311Packages.jaxlibWithoutCuda.dist)</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.jaxlibWithCuda.dist</li>
    <li>python311Packages.jaxopt</li>
    <li>python311Packages.jaxopt.dist</li>
    <li>python311Packages.jmp</li>
    <li>python311Packages.jmp.dist</li>
    <li>python311Packages.mplhep</li>
    <li>python311Packages.mplhep.dist</li>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
    <li>python311Packages.optax</li>
    <li>python311Packages.optax.dist</li>
    <li>python311Packages.optax.testsout</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
    <li>python311Packages.pytensor</li>
    <li>python311Packages.pytensor.dist</li>
    <li>python311Packages.skrl</li>
    <li>python311Packages.skrl.dist</li>
    <li>python311Packages.tensorflow-bin</li>
    <li>python311Packages.tensorflow-bin.dist</li>
    <li>python311Packages.torchrl</li>
    <li>python311Packages.torchrl.dist</li>
    <li>python311Packages.translatehtml</li>
    <li>python311Packages.translatehtml.dist</li>
    <li>python311Packages.uproot</li>
    <li>python311Packages.uproot.dist</li>
    <li>python311Packages.vector</li>
    <li>python311Packages.vector.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
    <li>wyoming-faster-whisper</li>
    <li>wyoming-faster-whisper.dist</li>
  </ul>
</details>

samuela committed:
python3Packages.jaxlibWithCuda: fix #282184 and refactor

This change snowballed to cover a number of improvements:
1. Fix #282184. The `--config=cuda` flag was lost at some point, disabling CUDA builds even with `cudaSupport = true`.
2. Migrate to using the CUDA redist packages instead of cudatoolkit.
3. Unify stdenv behind `effectiveStdenv` following a pattern that has precedent in OpenCV's derivation and was recommended in https://github.com/NixOS/nixpkgs/pull/288857#discussion_r1490003016.

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -65,25 +71,51 @@ let
+    name = "cuda-cc-joined";
     paths = [
-      backendStdenv.cc
+      effectiveStdenv.cc
       binutils.bintools # for ar, dwp, nm, objcopy, objdump, strip
```
Just for awareness (very much out of scope and potentially non-trivial to fix) `binutils` here and `cuda_nvcc.bin` above really should be from `buildPackages`. Our best prospect in the near future is splicing. Since this includes some refactors already, maybe we could mark all of the actually-native tools with a comment.

The motivation is that for somebody not working with cuda actively it's not going to be obvious what cuda_nvcc even is. I suffer from this issue with ROCm actually

samuela replied:
Ah, thanks for the pointer @SomeoneSerge ! I propose that we leave that to a future PR in the interest of keeping PRs as small as possible.

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -204,50 +270,80 @@ let
+      ''
+        cat <<CFG > ./.jax_configure.bazelrc
+        build --strategy=Genrule=standalone
+        build --repo_env PYTHON_BIN_PATH="${python}/bin/python"
```
I think this one is also native/executed at build time... although it's not like cross-compilation currently works with python packages at all

SomeoneSerge reviewed (approved):
I mostly just skimmed, but I think this should be good to go (my comments are nits)

samuela committed:
python3Packages.jaxlib: `--set-rpath` -> `--add-rpath`

samuela commented:
Thanks for the reviews @SomeoneSerge and @GaetanLepage! I'll proceed to merge tomorrow unless there are any objections?
